### PR TITLE
CI: Get coverage of filecheck tests

### PR DESCRIPTION
**Pull request by webmiche**
This PR leverages the [Coverage](https://coverage.readthedocs.io/en/7.0.0b1/api.html) package to generate the code coverage of our file checks test suite. Currently, there are a couple of choices that are controversial, which I would like to hear your perspectives. I will add comments to the corresponding locations in the file.

### CODE REVIEW

File: xdsl/tools/xdsl-opt
Comment (webmiche) [inline xdsl/tools/xdsl-opt:24]: Currently, the coverage is generated by default. This is a choice I made to keep the impact (and therefore the review time) as low as possible in this PR.
What we can also do is:
- only generate coverage if a flag is present (this will require adding this flag to all filecheck tests)
- generate by default, but not generate if if a flag is present
- always generate the coverage (the current version)

Furthermore, notice that these changes are in the `tools` directory and not in the big `xDSLOptMain` class. This is another choice we have. We can either leave them here, therefore not making users include this flag when building their own `...-opt` tools, or we can move it to the big class. Both are fine for me.

File: xdsl/tools/xdsl-opt
Comment (webmiche) [inline xdsl/tools/xdsl-opt:29]: This is a very weird issue and comes from the fact that `coverage` is launched in different subdirectories when using lit. This means that `.coveragerc` is sometimes not found or in different relative paths. Does anyone have a cleaner idea how to implement this?

File: xdsl/tools/xdsl-opt
Comment (webmiche) [inline xdsl/tools/xdsl-opt:44]: This will require further investigation, which I plan to work on today. If anyone has an idea/experience already, feel free to comment!

File: xdsl/tools/xdsl-opt
Comment (webmiche) [inline xdsl/tools/xdsl-opt:38]: This argument should make it s.t. the coverage is only reported for files within the `xdsl/` directory. To me, this makes sense as it is the thing we want to test. However, it is different to what the configuration for pytest currently does (can be found in `.coveragerc`). Is there a reason for that? Maybe @georgebisbas

I will look into synchronizing the two s.t. these arguments can be loaded form the `.coveragerc` directly.

File: xdsl/tools/xdsl-opt
Comment (webmiche) [inline xdsl/tools/xdsl-opt:55]: I also need to look into a way to not generate this report each time `xdsl-opt` is executed.

**Comment by webmiche** [comment]
Notice that this currently does not include the coverage of tests that depend on the MLIR Pyhton bindings as the coverage is uploaded for the Python-based testing only.

File: xdsl/tools/xdsl-opt
Comment (webmiche) [reply xdsl/tools/xdsl-opt:38]: The fact that the covered files is reduced in this PR might be the reason why the test coverage drops in the PR.

Comment (webmiche) [reply xdsl/tools/xdsl-opt:38]: Or it might be the case that through specifying a source, coverage can now understand that some of the files are not actually being executed.

Comment (webmiche) [reply xdsl/tools/xdsl-opt:38]: Yes, it seems that this PR covers 37 files, whereas the old coverage data only covered 29 files.

File: xdsl/tools/xdsl-opt
Comment (webmiche) [reply xdsl/tools/xdsl-opt]: This seems to be a race condition in Coverage related: https://github.com/nedbat/coveragepy/issues/1303

I rewrote the code s.t. each test generates its own coverage file, which get combined afterward by the coverage package. This way, no race condition should occur as distinct files are accessed.

File: xdsl/tools/xdsl-opt
Comment (webmiche) [reply xdsl/tools/xdsl-opt]: I have synchronized the two ways of generating coverage data. Now, only the `xdsl` folder is considered, which seems to hurt our coverage further.

File: xdsl/tools/xdsl-opt
Comment (webmiche) [reply xdsl/tools/xdsl-opt]: Done (described above).

File: .github/workflows/ci-core.yml
Comment (georgebisbas) [inline .github/workflows/ci-core.yml:37]: Why are you doing this change?

File: xdsl/tools/xdsl-opt
Comment (math-fehr) [reply xdsl/tools/xdsl-opt:29]: Can we give as option to `xdsl-opt` the link to `.coveragerc` ?

Comment (webmiche) [reply xdsl/tools/xdsl-opt:29]: The problem is that this cannot be an absolute path, since those are different on different machines, but the relative paths are always different since the program runs in different directories in each lit run.

Another potential solution is trying to configure lit in a way that it actually executes everything in the root directory. Then this would be unnecessary as well.

File: tests/filecheck/lit.cfg
Comment (webmiche) [inline tests/filecheck/lit.cfg:8]: Maybe I am missing something. From the [docu](https://llvm.org/docs/CommandGuide/lit.html#test-suites):
```
test_exec_root: For out-of-dir builds, the path to the test suite root inside the object directory. This is where tests will be run and temporary output files placed.
```

This is all nice and good for the tests defined in `tests/filecheck/`. The coverage files for those are generated in the expected `xdsl` folder. However, for every subfolder, say `tests/filecheck/irdl-printer/`, the coverage files are generate in a *new* folder `xdsl/irdl-printer`. Why does this happen? If I print the current working directory from within `xdsl-opt` I can clearly see that it is not executed in the `test_exec_root` but in a new directory `test_exec_root/*whatever is after tests/filecheck*`.

I will give up for today. I'll try to look into it again tomorrow and start to find workaround otherwise...

Comment (math-fehr) [reply tests/filecheck/lit.cfg:8]: So you mean that there is a `filecheck` that is appended at the end? That is weird :/ Maybe a bug in lit?

Comment (webmiche) [reply tests/filecheck/lit.cfg:8]: Not quite, there are subdirectories in the `filecheck` folder created and appended.

Anywho, I will try to workaround this. It will probably result in the pytest config and the lit config not being at the same place, which is unfortunate, but probably okay for a first version. I will add a TODO though.

File: .github/workflows/ci-core.yml
Comment (webmiche) [reply .github/workflows/ci-core.yml]: Reverted this.

File: tests/filecheck/lit.cfg
Comment (webmiche) [reply tests/filecheck/lit.cfg]: I have now split the configs again. Therefore, lit does not have to execute in the proper folder and I can collect the coverage data in the combine call.

I'll look into defining aliases in lit. I think it would be quite lit (badumtss) if we could give it a flag that generates coverage. Otherwise, running tests locally will create a ton of unnecessary files.

**Comment by webmiche** [comment]
Okay, I think I have a design that somehwat makes sense now. Considering that lit seems to be weird around the `exec_root` thing, I decided to dodge it. I have instead defined two user flags for lit:
- "-DCOVERAGE" makes filecheck tests generate coverage data
- "-DCOVERAGE_CONFIG" can be passed to specify the coverage config (in our case `xdsl/.coveragerc`). I then set this using the absolut path from within the CI, therefore sidestepping the executiong root issues.

This also means that the coverage files get created next to the filecheck files. Therefore, the `coverage combine` command got a bit more involved, specifying all subfolders. If anyone sees a nicer way of doing this, tell me pls :)
I have look into the using `**`, but wasn't able to get something to work that takes arbitrarily deeply nested subdirectories into account.

The aforementioned flags are passed to xdsl-opt in two steps:
- firstly, xdsl-opt now has two arguments `--generate-coverage` and `--coverage-config[=PATH]`, which correspond basically one-to-one to the aforementioned two flags.
- Secondly, I use custom substitutions in lit in order to substitute `%cov` to either `` (no coverage flag), `--generate-coverage` (only `DCOVERAGE`) or `--generate-coverage --coverage-config=[PATH]` (both flags).

File: .github/workflows/ci-core.yml
Comment (webmiche) [inline .github/workflows/ci-core.yml:41]: IMO, we might want to think about getting the test coverage from the MLIR-based CI instead of this one. It only takes around 4minuts more to run it and we would get better coverage data. Additionally, it usually takes the codecov bot around 5minutes to respond to a change, so essentially we would increase the time the bot needs to around 10 minutes, which I feel is not that much worse than 5 minutes. What do you think?

Comment (math-fehr) [reply .github/workflows/ci-core.yml:41]: I think it's fine to wait for 10 minutes for coverage! Most of the time, it takes longer to review anyway ;)

Comment (webmiche) [reply .github/workflows/ci-core.yml:41]: Moved it :+1:

**Review by math-fehr — COMMENTED** [review]
Looks really nice now!
I added some comments, to see if we can make this a bit less error-prone for when we will update it, but that's really cool!

File: tests/filecheck/lit.cfg
Comment (math-fehr) [reply tests/filecheck/lit.cfg]: Oh yeah, I haven't thought of that ^^'

File: .github/workflows/ci-core.yml
Comment (webmiche) [reply .github/workflows/ci-core.yml]: Changed to `coverage combine --append $(find tests/filecheck -type d)`.

File: tests/filecheck/lit.cfg
Comment (webmiche) [reply tests/filecheck/lit.cfg]: I guess it seems to work, so all good :)

**Review by math-fehr — APPROVED** [review]
Looks good to me!
Should we merge this now, or do you still want to add something else to it?

**Comment by webmiche** [comment]
Yes, as somewhat expected/feared the .xml file was not found to upload. I will look into this on my way home, hoping I can fix it in time. O/w, we might have to wait until monday to merge this.

If I succeed, I will just merge :)

**Comment by webmiche** [comment]
This is extremely weird. If I hard code the path to the file, it works. However, if I try to use the environment variable (so how a sane person would do it), the file is not found. Currently, I am trying to not pass a path and hope it works magically.

If not, I will probably look into this again on Monday and merge it then...

**Comment by webmiche** [comment]
Well, it found everything, only the one it should have is missing...

**Comment by math-fehr** [comment]
Thanks a lot for all this work. Personally, I don't really know how these things work, but I can take a look if you want!

**Comment by webmiche** [comment]
I think I am getting closer. Now, the file is discovered, but it seems that some of the tests are not taken into account. That's why the coverage drops by so much...

**Comment by webmiche** [comment]
Noting for my future self: There are two issues atm: Both lit and pytest act like they didn't have the Python Bindings installed. Additionally, the ${GITHUB_WORKSPACE} env variable seems weird, and pytest hsa the wrong root directory.

File: .github/workflows/ci-mlir.yml
Comment (georgebisbas) [inline .github/workflows/ci-mlir.yml:78]: minor typo

**Comment by webmiche** [comment]
Okay, I think this is about to get finished. I had one run earlier that behaved the way I wanted it too. Notice that we now lose some coverage as I have specified the source directories, so non-executed files are counted as 0% coverage. Furthermore, coverage is now generated on the MLIR CI.

Sidethought: We might want to think about what to do with all the `__init__` files, as they count towards coverage now, but actually aren't possible to cover 100% within one CI run.

**Pull request merged**
